x/twap/logic_test.go

```diff
@@ -521,3 +522,422 @@ func (s *TestSuite) TestPruneRecords() {
+					lastErrorTime: baseRecord.Time,
+				},
+				// The new record added.
+				{
+					spotPriceA:    sdk.OneDec(),
+					spotPriceB:    sdk.ZeroDec(),
+					lastErrorTime: baseRecord.Time.Add(time.Second), // equals to block time
+				},
+			},
+		},
+		"two-asset; pre-set at t and t + 1, new record with updated spot price created": {
+			preSetRecords: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			poolId:        baseRecord.PoolId,
+
+			blockTime: baseRecord.Time.Add(time.Second * 11),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.OneDec(),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.OneDec().Add(sdk.OneDec()),
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA: sdk.OneDec(),
+					spotPriceB: sdk.OneDec().Add(sdk.OneDec()),
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record at t.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The original record at t + 1.
+				{
+					spotPriceA: tPlus10sp5Record.P0LastSpotPrice,
+					spotPriceB: tPlus10sp5Record.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA: sdk.OneDec(),
+					spotPriceB: sdk.OneDec().Add(sdk.OneDec()),
+				},
+			},
+		},
+		// This case should never happen in-practice since ctx.BlockTime
+		// should always be greater than the last record's time.
+		"two-asset; pre-set at t and t + 1, new record inserted between existing": {
+			preSetRecords: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			poolId:        baseRecord.PoolId,
+
+			blockTime: baseRecord.Time.Add(time.Second * 5),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.OneDec(),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.OneDec().Add(sdk.OneDec()),
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA: sdk.OneDec(),
+					spotPriceB: sdk.OneDec().Add(sdk.OneDec()),
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record at t.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				// TODO: it should not be possible to add a record between existing records.
+				// https://github.com/osmosis-labs/osmosis/issues/2686
+				{
+					spotPriceA: sdk.OneDec(),
+					spotPriceB: sdk.OneDec().Add(sdk.OneDec()),
+				},
+				// The original record at t + 1.
+				{
+					spotPriceA: tPlus10sp5Record.P0LastSpotPrice,
+					spotPriceB: tPlus10sp5Record.P1LastSpotPrice,
+				},
+			},
+		},
+		// TODO: complete multi-asset pool tests:
```

Comment:
let's create an issue for this so that we come back to this!

Reply:
Already made: https://github.com/osmosis-labs/osmosis/issues/2688